### PR TITLE
[AAE-4966] Extensible app config

### PR DESCRIPTION
**Pull request by DenysVuika**
**Please check if the PR fulfills these requirements**

> - [x] The commit message follows our [guidelines](https://github.com/Alfresco/alfresco-ng2-components/wiki/Commit-format)
> - [x] Tests for the changes have been added (for bug fixes / features)
> - [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

> - [ ] Bugfix
> - [x] Feature
> - [ ] Code style update (formatting, local variables)
> - [ ] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:

**What is the current behaviour?** (You can also link to an open issue here)

**What is the new behaviour?**

- support for updating App Config from the extensions ("appConfig" property)
- cleanup the incorrect usage of the service injection in ADF unit tests (stability and less refactorings in the future)

**Does this PR introduce a breaking change?** (check one with "x")

> - [ ] Yes
> - [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
https://alfresco.atlassian.net/browse/AAE-4966

### CODE REVIEW

File: lib/extensions/src/lib/services/extension.service.ts [hunk 1]
```diff
@@ -58,6 +59,8 @@ export class ExtensionService {
     features: Array<any> = [];
     authGuards: { [key: string]: Type<{}> } = {};
 
+    readonly setup$ = new BehaviorSubject<ExtensionConfig>(this.config);
```
Comment (popovicsandras) [inline lib/extensions/src/lib/services/extension.service.ts:62]: Can you please reveal the Observable interface only here? I don't think making it readonly helps us here, someone from outside can still call `next()` on this BehaviourSubject.

**Review by popovicsandras — CHANGES_REQUESTED** [review]

File: lib/extensions/src/lib/services/extension.service.ts (hunk 1 above)
Comment (DenysVuika) [reply lib/extensions/src/lib/services/extension.service.ts:62]: how should a unit test interact with these events? replacing the entire service instance with a custom one?

**Comment by mauriziovitale** [comment]
Do we have an issue linked that contains the AC? Once merged, who is gonna test it and based on what? We should have examples and expectations
fileExtension1
fileExtension2

result

File: lib/extensions/src/lib/services/extension.service.ts (hunk 1 above)
Comment (DenysVuika) [reply lib/extensions/src/lib/services/extension.service.ts]: @popovicsandras can you please check the updated code

**Comment by DenysVuika** [comment]
This code is needed for https://alfresco.atlassian.net/browse/AAE-4842. The service has unit tests, the corresponding UI in the Modeling app will be testable.

File: lib/extensions/src/lib/services/extension.service.ts (hunk 1 above)
Comment (popovicsandras) [reply lib/extensions/src/lib/services/extension.service.ts]: A common technique for this is that for the sake of unit test, you can inject a very tiny factory function as the last parameter of the constructor:
```
function defaultSubjectFactory<T>(initParam) {
    return new BehaviorSubject<T>(initParam);
}

@Injectable({providedIn: 'root'})
export class ExtensionService {

    // ...

    private setupSubject$: BehaviorSubject<ExtensionConfig>;
    setup$: Observable<ExtensionConfig>;

    constructor(/* ...every other constructor parameter, */ subjectFactory = defaultSubjectFactory) {
        // ...
        this.setupSubject$ = subjectFactory<ExtensionConfig>(this.config);
        this.setup$ = this.setupSubject$.asObservable();
        // ...
    }

    // ...
}
```

Comment (DenysVuika) [reply lib/extensions/src/lib/services/extension.service.ts]: solved with protected member and custom service mock for unit tests

File: lib/core/app-config/app-config.service.ts [hunk 2]
```diff
@@ -69,9 +70,19 @@ export class AppConfigService {
     protected onLoadSubject: Subject<any>;
     onLoad: Observable<any>;
 
-    constructor(private http: HttpClient) {
+    constructor(private http: HttpClient, extensionService: ExtensionService) {
         this.onLoadSubject = new Subject();
         this.onLoad = this.onLoadSubject.asObservable();
+
+        extensionService.setup$.subscribe((config) => {
```
[the remark this answers is not in the document]
Comment (DenysVuika) [reply lib/core/app-config/app-config.service.ts:77]: Updated the code to re-merge extensions in this case

File: lib/extensions/src/lib/config/extension.config.ts [hunk 3]
```diff
@@ -27,9 +27,9 @@ export interface ExtensionRef {
     $license: string;
     $runtime?: string;
     $description?: string;
-
     $dependencies?: Array<string>;
 
+    appConfig?: any;
```
Comment (pmartinezga) [inline lib/extensions/src/lib/config/extension.config.ts:32]: Shold we update the json schema for the extension files?

**Review by pmartinezga — COMMENTED** [review]
Minor comment

**Review by mauriziovitale — APPROVED** [review]

File: lib/extensions/src/lib/config/extension.config.ts (hunk 3 above)
Comment (DenysVuika) [reply lib/extensions/src/lib/config/extension.config.ts:32]: @pmartinezga we will have to maintain the whole app.config schema inside the extension schema I'm afraid

Comment (pmartinezga) [reply lib/extensions/src/lib/config/extension.config.ts:32]: won't it be enough by adding the appConfig attribute as an object? Something like adding here a piece of json like this:
```
 "$appConfig": {
      "description": "Application configuration parameters to add/override",
      "type": "object"
}
```

Comment (DenysVuika) [reply lib/extensions/src/lib/config/extension.config.ts:32]: updated schema, thanks!

**Review by pmartinezga — APPROVED** [review]

**Pull request merged**